src/blog/understanding-ecmascript-part-3.md

```diff
@@ -0,0 +1,362 @@
+
+> [<code>BindingIdentifier<sub>[Yield, Await]</sub> :</code>](https://tc39.es/ecma262/#prod-BindingIdentifier)
+> `Identifier`
+> `yield`
+> `await`
+
+Expanded, this means the following productions:
+
+> `BindingIdentifier_Await :`
+> `Identifier`
+> `yield`
+> `await`
+>
+> `BindingIdentifier :`
+> `Identifier`
+> `yield`
+> `await`
+
+(We're omitting the productions for `BindingIdentifier_Yield` and `BindingIdentifier_Yield_Await` which are not needed in our example.)
+
+This looks like `await` and `yield` would be always allowed as identifiers. What's up with that? Is the whole blog post for nothing?
+
+### Statics semantics to the rescue
+
+Turns out **static semantics** are needed for forbidding `await` as an identifier inside async functions.
+
+Static semantics describe static rules &mdash; that is, rules that are checked before the program runs.
+
+In this case, the [static semantics for BindingIdentifier](https://tc39.es/ecma262/#sec-identifiers-static-semantics-early-errors) define the following syntax-directed rule:
+
+>`BindingIdentifier : await`
+>
+> It's a Syntax Error if this production has an <code><sub>[Await]</sub></code> parameter.
+
+Effectively, this forbids the `BindingIdentifier_Await : await` production.
+
+The spec explains that the reason for having this production but defining it as a Syntax Error by the static semantics is because of interference with automatic semicolon insertion (ASI).
+
+Remember that ASI kicks in when we're unable to parse a line of code according to the grammar productions. ASI tries to add semicolons to satisfy the requirement that statements and declarations must end with a semicolon. (We'll describe ASI in more detail in a later episode.)
+
+Consider the following code (example from the spec):
+
+```js
+async function too_few_semicolons() {
+    let
+    await 0;
+}
+```
+
+If the grammar disallowed `await` as an identifier, ASI would kick in and transform the code into the following grammatically correct code, which also uses `let` as an identifier:
+
+```js
+async function too_few_semicolons() {
+    let;
+    await 0;
+}
+```
+
+This kind of inference with ASI was deemed too confusing, so static semantics were used for disallowing `await` as an identifier.
+
+### Disallowed `StringValues` of identifiers
+
+There's also another related rule:
+
+> `BindingIdentifier : Identifier`
+>
+> It is a Syntax Error if this production has an <code><sub>[Await]</sub></code> parameter and `StringValue` of `Identifier` is `"await"`.
+
+This might be confusing at first. `Identifier` is defined like this:
+
+> [`Identifier :`](https://tc39.es/ecma262/#prod-Identifier)
+> `IdentifierName` but not `ReservedWord`
+
+`await` is a `ReservedWord`, so how can an `Identifier` ever be `await`?
+
+Turns out, `Identifier` cannot be `await`, but it can be something else whose `StringValue` is `"await"` &mdash; a different representation of the character sequence `await`.
```

Comment:
```suggestion
As it turns out, `Identifier` cannot be `await`, but it can be something else whose `StringValue` is `"await"` &mdash; a different representation of the character sequence `await`.
```